### PR TITLE
Drools 4268 - FIX XStream configuration

### PR DESCRIPTION
**Pull request by gitgabrio**
@kkufova @danielezonca @yesamer 

See https://issues.jboss.org/browse/DROOLS-4268

### CODE REVIEW

File: drools-scenario-simulation/drools-scenario-simulation-api/src/main/resources/org/drools/scenariosimulation/scesim.xsd [hunk 1]
```diff
@@ -0,0 +1,126 @@
+<xs:schema attributeFormDefault="unqualified" elementFormDefault="qualified" xmlns:xs="http://www.w3.org/2001/XMLSchema">
```
Comment (danielezonca) [inline drools-scenario-simulation/drools-scenario-simulation-api/src/main/resources/org/drools/scenariosimulation/scesim.xsd]: XSD introduction is related to client side marshalling that is not part of this PR so I think should not be added here. Wdyt?
If you decide to keep it please remove commented nodes

File: drools-scenario-simulation/drools-scenario-simulation-backend/src/main/java/org/drools/scenariosimulation/backend/util/ScenarioSimulationXMLPersistence.java [hunk 2]
```diff
@@ -37,35 +40,60 @@
...
 
+    public static String getCurrentVersion() {
+        return currentVersion;
+    }
+
+    public static String cleanUpUnusedNodes(String input) {
+        return removeNode(input, "<simulationDescriptor reference\\b[^>]*>(.*?)");
+    }
+
+    private static  String removeNode(String input, String toRemove) {
```
Comment (danielezonca) [inline drools-scenario-simulation/drools-scenario-simulation-backend/src/main/java/org/drools/scenariosimulation/backend/util/ScenarioSimulationXMLPersistence.java]: This is not a "remove node" but just a string replace, I think the name is a bit misleading. Wdyt?

File: drools-scenario-simulation/drools-scenario-simulation-backend/src/main/java/org/drools/scenariosimulation/backend/util/InMemoryMigrationStrategy.java [hunk 3]
```diff
@@ -113,4 +127,30 @@
...
+            // Unmarshall the 1.5 format with "older" xstream configuration, that read "reference" attributes
+            Object o = getLocalXStream().fromXML(input);
+            ScenarioSimulationModel model = (ScenarioSimulationModel) o;
+
+            // Marshall the model with the "new" xstream configuration, that does not write "reference" attributes
+            return updateVersion(xmlPersistence.marshal(model), "1.5", "1.6");
+        };
+    }
+
+    private XStream getLocalXStream() {
```
Comment (danielezonca) [inline drools-scenario-simulation/drools-scenario-simulation-backend/src/main/java/org/drools/scenariosimulation/backend/util/InMemoryMigrationStrategy.java:145]: This method should be used only with files version <= 1.5. Add please a java doc or rename the method

File: drools-scenario-simulation/drools-scenario-simulation-backend/src/main/java/org/drools/scenariosimulation/backend/util/ScenarioSimulationXMLPersistence.java [hunk 4]
```diff
@@ -37,35 +40,60 @@
...
-        xt.alias("SimulationDescriptor", SimulationDescriptor.class);
-
-        xt.alias("Import", Import.class);
+    /**
+     * Method to configure the commonly-used mappings defined in {@see XSTREAM_MAPPINGS}
+     * @param toConfigure
+     */
+    public static void configureXStreamMappings(XStream toConfigure) {
+        Map<String, Class> XSTREAM_MAPPINGS = new HashMap<String, Class>() {
+            {
```
Comment (danielezonca) [inline drools-scenario-simulation/drools-scenario-simulation-backend/src/main/java/org/drools/scenariosimulation/backend/util/ScenarioSimulationXMLPersistence.java]: Using this syntax you are creating a subclass of `HashMap` everytime this method is invoked. Can you please refactor to avoid the creation of this map?

**Review by danielezonca — CHANGES_REQUESTED** [review]

**Comment by baldimir** [comment]
jenkins execute compile downstream build

File: drools-scenario-simulation/drools-scenario-simulation-api/src/main/resources/org/drools/scenariosimulation/scesim.xsd (hunk 1 above)
Comment (gitgabrio) [reply drools-scenario-simulation/drools-scenario-simulation-api/src/main/resources/org/drools/scenariosimulation/scesim.xsd]: @danielezonca 
Yes - it currently is not part of this PR. Removed

File: drools-scenario-simulation/drools-scenario-simulation-backend/src/main/java/org/drools/scenariosimulation/backend/util/ScenarioSimulationXMLPersistence.java (hunk 2 above)
Comment (gitgabrio) [reply drools-scenario-simulation/drools-scenario-simulation-backend/src/main/java/org/drools/scenariosimulation/backend/util/ScenarioSimulationXMLPersistence.java]: @danielezonca 
Removed

File: drools-scenario-simulation/drools-scenario-simulation-backend/src/main/java/org/drools/scenariosimulation/backend/util/ScenarioSimulationXMLPersistence.java (hunk 4 above)
Comment (gitgabrio) [reply drools-scenario-simulation/drools-scenario-simulation-backend/src/main/java/org/drools/scenariosimulation/backend/util/ScenarioSimulationXMLPersistence.java]: @danielezonca 
Fixed

**Comment by gitgabrio** [comment]
jenkins please retest this

File: drools-scenario-simulation/drools-scenario-simulation-api/src/main/java/org/drools/scenariosimulation/api/model/Scenario.java [hunk 5]
```diff
@@ -37,10 +35,6 @@
     private final List<FactMappingValue> factMappingValues = new ArrayList<>();
 
 
-    public Scenario() {
```
Comment (danielezonca) [inline drools-scenario-simulation/drools-scenario-simulation-api/src/main/java/org/drools/scenariosimulation/api/model/Scenario.java:43]: Empty constructor is needed for CDI so just removing it could be a bit error prone in the future (we need to remember that as soon as we will need another constructor we will need to add back also the empty one).
What about a javadoc note?

File: drools-scenario-simulation/drools-scenario-simulation-backend/src/main/java/org/drools/scenariosimulation/backend/util/ScenarioSimulationXMLPersistence.java [hunk 6]
```diff
@@ -134,6 +123,9 @@ public String migrateIfNecessary(String rawXml) {
             case "1.5":
                 migrator = migrator.andThen(getMigrationStrategy().from1_5to1_6());
                 supported = true;
+                break;
+            default:
+                supported = currentVersion.equals(fileVersion);
```
Comment (danielezonca) [inline drools-scenario-simulation/drools-scenario-simulation-backend/src/main/java/org/drools/scenariosimulation/backend/util/ScenarioSimulationXMLPersistence.java:128]: Sonar complains with `case` without `break` but I don't like to add it only for 1.5 (Sonar is analyzing only new code). I would like to refactor the whole switch (to `if/else` probably) or keep last case without and we will do in another ticket. Wdyt?

**Review by danielezonca — CHANGES_REQUESTED** [review]
Additional two comments

File: drools-scenario-simulation/drools-scenario-simulation-backend/src/main/java/org/drools/scenariosimulation/backend/util/ScenarioSimulationXMLPersistence.java (hunk 6 above)
Comment (gitgabrio) [reply drools-scenario-simulation/drools-scenario-simulation-backend/src/main/java/org/drools/scenariosimulation/backend/util/ScenarioSimulationXMLPersistence.java:128]: @danielezonca 
I am not really sure about the reason behind that implementation - but changing the logic may require more tests and is out of scope of this PR. What is clearly not-so-nice is the mix of two different approaches: switch/case is meant exactly to provide a "default" value if specific "cases" are not executed. But in the original implementation the "default" value was provided before the switch, and the switch' default was ignored. IMHO for now this fix it is just a readability improvement; later eventually we can dig more on the execution logic behind.

File: drools-scenario-simulation/drools-scenario-simulation-api/src/main/java/org/drools/scenariosimulation/api/model/Scenario.java (hunk 5 above)
Comment (gitgabrio) [reply drools-scenario-simulation/drools-scenario-simulation-api/src/main/java/org/drools/scenariosimulation/api/model/Scenario.java:43]: @danielezonca 
I may be wrong, but empty constructor is needed only when there is another (not empty) one: it seems also Simulation and SImulationDescriptor does not explicitly declare it

Comment (danielezonca) [reply drools-scenario-simulation/drools-scenario-simulation-api/src/main/java/org/drools/scenariosimulation/api/model/Scenario.java:43]: Yes, I meant that as soon as we will need another constructor we need to remember to add back also the empty constructor because of CDI. So I was suggesting a note in javadoc but just a comment not really a request for change

File: drools-scenario-simulation/drools-scenario-simulation-backend/src/main/java/org/drools/scenariosimulation/backend/util/ScenarioSimulationXMLPersistence.java (hunk 6 above)
Comment (danielezonca) [reply drools-scenario-simulation/drools-scenario-simulation-backend/src/main/java/org/drools/scenariosimulation/backend/util/ScenarioSimulationXMLPersistence.java:128]: Fine for me

**Review by danielezonca — APPROVED** [review]

File: drools-scenario-simulation/drools-scenario-simulation-api/src/main/java/org/drools/scenariosimulation/api/model/Scenario.java [hunk 7]
```diff
@@ -118,7 +106,7 @@ public void resetErrors() {
     }
 
     Scenario cloneScenario() {
-        Scenario cloned = new Scenario(simulationDescriptor);
+        Scenario cloned = new Scenario();
```
Comment (yesamer) [inline drools-scenario-simulation/drools-scenario-simulation-api/src/main/java/org/drools/scenariosimulation/api/model/Scenario.java:109]: In my opinion this method is now misleading: before the change, method 'cloneScenario()' creates a new Scenario instance with the same scenarioDescriptor in scope. Now, it simpy creates a totally different instance of Scenario. Therefore, if the method is still used, it should be renamed to 'createScenario()' or it could be removed, and create a new scenario instance directly in the caller.

File: drools-scenario-simulation/drools-scenario-simulation-backend/src/main/java/org/drools/scenariosimulation/backend/util/MigrationStrategy.java [hunk 8]
```diff
@@ -61,6 +61,12 @@
      */
     Function<String, String> from1_4to1_5();
 
+    /**
+     * Method to obtain the migration function from 1.4 to 1.5
```
Comment (yesamer) [inline drools-scenario-simulation/drools-scenario-simulation-backend/src/main/java/org/drools/scenariosimulation/backend/util/MigrationStrategy.java]: Probably you mean from 1.5 to 1.6 in this JavaDoc

**Review by yesamer — CHANGES_REQUESTED** [review]

File: drools-scenario-simulation/drools-scenario-simulation-api/src/main/java/org/drools/scenariosimulation/api/model/Scenario.java (hunk 7 above)
Comment (gitgabrio) [reply drools-scenario-simulation/drools-scenario-simulation-api/src/main/java/org/drools/scenariosimulation/api/model/Scenario.java:109]: @yesamer 
I don't think so. "SimulationDescriptor" was a leftover unused anywhere. The main point is the cloning of all fact mappings - that's why it is different from a simple "createScenario" - for which, a simple new Scenario() is enough. Beside that, the meant semantic is: create an exact copy of the original: whatever property will be add/removed from Scenario, it is clear it will have to be managed here.

**Review by yesamer — APPROVED** [review]

**Review by kkufova — APPROVED** [review]
Looks good to me.

:+1: 

Approving.

**Comment by gitgabrio** [comment]
jenkins please retest this

**Comment by danielezonca** [comment]
jenkins execute compile downstream build
